components/in_memory_engine/src/write_batch.rs

```diff
@@ -66,6 +67,14 @@ pub struct RegionCacheWriteBatch {
 
     // record the total durations of the prepare work for write in the write batch
     prepare_for_write_duration: Duration,
+
+    // Now, we have an assumption that in one round of batch system process, although the same
+    // region can call `prepare_for_region` multiple times, it can only call sequentially. This is
+    // say, we will not have this: prepare_for_region(region1), prepare_for_region(region2),
+    // prepare_for_region(region1).  In case to avoid this asssumption being broken, we record the
+    // regions that have called prepare_for_region and ensure that if the region is not the
+    // `currnet_region`, it is not recorded in this vec.
```

Comment:
Could you elaborate the definition of "one round of batch system process"?

---

I'm not sure if this is relevant, but an apply fsm may be put back to "batch" immediately and poll (`handle_normal`) again if new messages arrives after `handle_nromal` and before `schedule`.

https://github.com/tikv/tikv/blob/0b1124b92896b48fdd8290a20e55246cf1299fe3/components/batch-system/src/batch.rs#L408-L447

Reply:
Im not sure what you mean of "poll again if new messages arrives".